util/src/main/java/org/killbill/billing/util/entity/dao/EntitySqlDaoWrapperInvocationHandler.java

```diff
@@ -236,11 +241,79 @@ public Object execute() throws Throwable {
             return obj;
         }
 
-        // PERF: override the return value with the reHydrated entity to avoid an extra 'get' in the transaction,
-        // (see EntityDaoBase#createAndRefresh for an example, but it works for updates as well).
-        final List<M> ms = updateHistoryAndAudit(entityIds, deletedEntities, changeType, context);
-        final boolean isBatchQuery = method.getAnnotation(SqlBatch.class) != null;
-        return isBatchQuery ? ms : Iterables.<M>getFirst(ms, null);
+        InternalCallContext context = null;
+        // Retrieve record_id(s) for audit and history tables
+        final List<Long> entityRecordIds = new LinkedList<Long>();
+        if (changeType == ChangeType.DELETE) {
```

Comment:
Am i understanding well, that all this new code is because we to optimize the rehydration process by using `getByRecordId` instead of `getBydId`?

Reply:
Not quite: the gist of it is that we have a matrix of cases `tableName` x `changeType` x `with or without history`. In some cases, we only need the new `record_id`, in others we need the old `entity` and the `new entity`, etc.

The previous code was a bit more generic and wasn't explicitly enumerating these matrix cases, causing extra unnecessary `get` queries to be issued (IIRC, I saw 2 unnecessary extra `getInvoiceItem` queries for each `createInvoiceItem`). I now go in each of these cases to try to optimize as much as possible. Code is a bit more verbose but more obvious IMHO (and I've added `Preconditions` to validate some of my assumptions).